File: LeadToOffer/37_NumInSortArray.cpp
//1. binary count. O(logn) time.
class Solution {
public:
    int binaryCount(vector<int> data, int k, int start, int end) {
        if (start>end)
            return 0;
        if (start==end){
            if (data[start]==k)
                return 1;
            else
                return 0;
        }
        int mid = (start+end)/2;
        if (data[mid]>k)
            return binaryCount(data, k, start, mid-1);
        else if (data[mid]<k)
            return binaryCount(data, k, mid+1, end);
        else
            return binaryCount(data, k, start, mid-1) + binaryCount(data, k, mid+1, end) + 1;
    }
    int GetNumberOfK(vector<int> data ,int k) {
        if (data.size()==0)
            return 0;
        int count = binaryCount(data, k, 0, data.size()-1);
        return count;
    }
};

File: Algo/LeadToOffer/37_NumInSortArray.py
class Solution:
    def GetNumberOfK(self , data: List[int], k: int) -> int:
        # write code here
        return binary_count(data, k, 0, len(data)-1)


def binary_count(data, k, l, r):
    if l > r:
        return 0
    if l == r:
        return 1 if data[l] == k else 0

    m = int((l + r) / 2)
    if data[m] == k:
        return binary_count(data, k, l, m-1) + binary_count(data, k, m+1, r) + 1
    elif data[m] > k:
        return binary_count(data, k, l, m-1)
    else:
        return binary_count(data, k, m+1, r)